File: trees_and_graphs/binary_trees/bfs/c++/level_averages/level_averages.cpp
// level_averages.cpp
#include "level_averages.hpp"
#include "tree_node.hpp"
#include <vector>
#include <variant>
#include <deque>

std::vector<double> computeLevelAverages(TreeNode::ptr &root)
{
    if (root == nullptr)
        return {};
    if (root->left == nullptr && root->right == nullptr)
    {
        int nodeData = std::get<int>(root->data);
        return {static_cast<double>(nodeData)};
    }

    std::deque<TreeNode *> queue = {root.get()};
    std::vector<double> output;

    while (!queue.empty())
    {
        const size_t numNodesCurrLevel = queue.size();
        long long int runningLevelSum = 0;

        for (size_t i = 0; i < numNodesCurrLevel; i++)
        {
            TreeNode *currNode = queue.front();
            queue.pop_front();

            if (std::holds_alternative<int>(currNode->data))
            {
                runningLevelSum += std::get<int>(currNode->data);
            }
            else
            {
                throw std::invalid_argument("TreeNode data should be int for this algorithm");
            }

            if (currNode->left)
                queue.push_back(currNode->left.get());
            if (currNode->right)
                queue.push_back(currNode->right.get());
        }

        output.push_back(static_cast<double>(runningLevelSum) / static_cast<double>(numNodesCurrLevel));
    }

    return output;
}

File: trees_and_graphs/binary_trees/bfs/python3/level_averages.py
from array import ArrayType, array
from collections import deque
from typing import Deque, List, Optional

from tree_node import TreeNode


def compute_level_averages(root: Optional[TreeNode]) -> ArrayType[float]:
    """
    Compute the average value of nodes at each level of a binary tree.

    Parameters
    ----------
    root : Optional[TreeNode]
        The root node of the binary tree.

    Returns
    -------
    ArrayType[float]
        An array of C doubles (8 bytes) where each element is the average
        of node values at the corresponding level in the tree.
    """
    if not root:
        return array("d")
    if not (root.left or root.right):
        assert isinstance(root.data, (int, float))
        return array("d", [float(root.data)])

    queue: Deque[TreeNode] = deque([root])
    output: ArrayType[float] = array("d")
    level: int = 0
    print(f"Initial state of the queue: {queue}")

    while queue:
        prefix: str = f"{level * ' '}Level {level} | "
        num_nodes_curr_level: int = len(queue)
        running_level_sum: float = 0
        print(f"{prefix}Number of nodes = {num_nodes_curr_level}")
        print(f"{prefix}Initial running sum = {running_level_sum}")

        for _ in range(num_nodes_curr_level):
            print(f"{prefix}State of queue: {queue}")
            curr_node: TreeNode = queue.popleft()
            print(f"{prefix}Dequeue and process {curr_node}")
            assert isinstance(curr_node.data, (int, float))
            print(
                f"{prefix}Running sum = {running_level_sum} + {curr_node.data} = {running_level_sum + curr_node.data}"
            )
            running_level_sum += curr_node.data

            if curr_node.left:
                queue.append(curr_node.left)
            if curr_node.right:
                queue.append(curr_node.right)
        level += 1
        output.append(running_level_sum / num_nodes_curr_level)

    return output


def main() -> int:
    test_cases: List[List[Optional[int]]] = [
        [3, 9, 20, None, None, 15, 7],
        [3, 9, 20, 15, 7],
        [],
        [2147483647, 2147483647, 2147483647],
    ]
    for node_data in test_cases:
        print(f"Case " + "-" * 60, "\n")
        root: Optional[TreeNode] = TreeNode.construct_binary_tree(values=node_data)
        output: ArrayType[float] = compute_level_averages(root=root)
        print(f"The array of level averages is: {output if output else []}\n")

    return 0


if __name__ == "__main__":
    main()
